CMakeLists.txt

```diff
@@ -75,18 +75,32 @@ IF(NOT KOKKOS_HAS_TRILINOS)
   cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
   set(CMAKE_DISABLE_SOURCE_CHANGES ON)
   set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
-  IF (Spack_WORKAROUND)
-    #if we are explicitly using Spack for development,
-    #nuke the Spack compiler
-    SET(SPACK_CXX $ENV{SPACK_CXX})
-    IF(SPACK_CXX)
-      SET(CMAKE_CXX_COMPILER ${SPACK_CXX} CACHE STRING "the C++ compiler" FORCE)
-      SET(ENV{CXX} ${SPACK_CXX})
+
+  # What language are we compiling Kokkos as
+  # Downstream dependencies need to match this!
+  SET(KOKKOS_CMAKE_COMPILE_LANGUAGE CXX)
+  IF (Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE)
+    SET(KOKKOS_INTERNAL_CMAKE_EXTRA_COMPILE_LANGUAGE CXX)
+    IF (Kokkos_ENABLE_CUDA)
+      SET(KOKKOS_CMAKE_COMPILE_LANGUAGE CUDA)
+    ENDIF()
+    IF (Kokkos_ENABLE_HIP)
+      SET(KOKKOS_CMAKE_COMPILE_LANGUAGE HIP)
+    ENDIF()
+  ELSE()
+    IF (Spack_WORKAROUND)
+      #if we are explicitly using Spack for development,
+      #nuke the Spack compiler
+      SET(SPACK_CXX $ENV{SPACK_CXX})
+      IF(SPACK_CXX)
+        SET(CMAKE_CXX_COMPILER ${SPACK_CXX} CACHE STRING "the C++ compiler" FORCE)
+        SET(ENV{CXX} ${SPACK_CXX})
+      ENDIF()
     ENDIF()
   ENDIF()
   # Always call the project command to define Kokkos_ variables
   # and to make sure that C++ is an enabled language
-  PROJECT(Kokkos CXX)
+  PROJECT(Kokkos ${KOKKOS_CMAKE_COMPILE_LANGUAGE} ${KOKKOS_INTERNAL_CMAKE_EXTRA_COMPILE_LANGUAGE})
```

Comment:
Because KOKKOS_CMAKE_COMPILE_LANGUAGE is what is getting used for all the cpp files etc. We still need CXX in here, so that the HOST compiler gets properly identified, also we use it to compile gtest I believe.

Reply:
If `KOKKOS_INTERNAL_CMAKE_EXTRA_COMPILE_LANGUAGE` is always `CXX` I don't see a reason for the extra variable as of now. Is it a problem to have `PROJECT(Kokkos CXX CXX)`?

Reply:
Actually I didn't test that, but that was the reason why I did it (the double CXX)

Reply:
Results in warning:
```
-- Detecting CXX compile features - done
CMake Warning (dev) at /net/watson.sandia.gov/storage/fast/projects/sems/install/rhel7-x86_64/sems/utility/cmake/3.17.1/share/cmake-3.17/Modules/CMakeTestCompilerCommon.cmake:13 (message):
  Ignored CHECK_PASS without CHECK_START
Call Stack (most recent call first):
  /net/watson.sandia.gov/storage/fast/projects/sems/install/rhel7-x86_64/sems/utility/cmake/3.17.1/share/cmake-3.17/Modules/CMakeTestCXXCompiler.cmake:59 (PrintTestCompilerResult)
  CMakeLists.txt:103 (PROJECT)
This warning is for project developers.  Use -Wno-dev to suppress it.
```

Reply:
Hmm... OK.